lib/shrine/plugins/derivation_endpoint.rb

```diff
@@ -0,0 +1,1295 @@
+    # ## Expiration
+    #
+    # By default generated URLs are valid indefinitely. If you want URLs to
+    # expire after a certain amount of time, you can set the `:expires_in`
+    # option:
+    #
+    #     plugin :derivation_endpoint, expires_in: 90
+    #
+    # Now any URL will stop being valid 90 seconds after it was generated:
+    #
+    #     uploaded_file.derivation_url(:thumbnail)
+    #     #=> "/thumbnail/eyJpZCI6ImZvbyIsInN?expires_at=1547843568&signature=..."
+    #
+    # You can also pass `:expires_in` per URL:
+    #
+    #     uploaded_file.derivation_url(:thumbnail, expires_in: 90)
+    #     #=> "/thumbnail/eyJpZCI6ImZvbyIsInN?expires_at=1547843568&signature=..."
+    #
+    # ## Response headers
+    #
+    # ### Content Type
+    #
+    # By default in the derivation response, the [`Content-Type`] header is
+    # inferred from the file extension of the derivative (using `Rack::Mime`).
+    # You can override that with the `:type` option:
+    #
+    #     plugin :derivation_endpoint, type: -> {
+    #       "image/webp" if name == :webp
+    #     }
+    #
+    # If the block returns `nil`, then the default type will be set. You can
+    # also set `:type` per URL:
+    #
+    #     uploaded_file.derivation_url(:webp, type: "image/webp")
+    #     #=> "/webp/eyJpZCI6ImZvbyIsInN?type=image%2Fwebp&signature=..."
+    #
+    # ### Content Disposition
+    #
+    # By default in the derivation response, the [`Content-Disposition`] header
+    # sets the disposition to `inline`, while the download filename is generated
+    # from derivation name, arguments and source file id. You can override that
+    # with the `:disposition` and `:filename` options:
+    #
+    #     plugin :derivation_endpoint,
+    #       disposition: -> { name == :thumbnail ? "inline" : "attachment" },
+    #       filename:    -> { [name, *args].join("-") }
+    #
+    # When the user opens the link in the browser, an `inline` disposition will
+    # tell the browser to render the file if possible, while `attachment`
+    # disposition will force download.
+    #
+    # The `:filename` and `:disposition` options can also be set per URL:
+    #
+    #     uploaded_file.derivation_url(:pdf, disposition: "attachment", filename: "custom-filename")
+    #     #=> "/thumbnail/eyJpZCI6ImZvbyIsInN?disposition=attachment&filename=custom-filename&signature=..."
+    #
+    # ## Uploading
+    #
+    # By default the derivation from a source file will be called each time
+    # it's requested. However, you can cache the derivatives to a storage by
+    # setting `:upload` to `true`. On first request the generated derivative
+    # will be uploaded to storage, and then on subsequent requests the already
+    # uploaded derivative will be served from the storage.
+    #
+    #     plugin :derivation_endpoint, upload: true
+    #
+    # The derivative will be uploaded to `<source id>/<name>-<args>` by default,
+    # and can be changed via `:upload_location`:
```

Comment:
This is in the ImageUploader `:store` storage right? And if I had 10 prefixes in that storage for performance, the derivative is uploaded to `/4_digit_storage_prefix/derivatives/source_id/name-args`.

Reply:
Yes, that's correct. You decide which prefixes you want for performance.